core/src/main/java/jenkins/agents/WebSocketAgents.java

```diff
@@ -166,11 +166,19 @@ protected void error(Throwable cause) {
 
         class Transport extends AbstractByteBufferCommandTransport {
 
+            Transport() {
+                super(true);
+            }
+
             @Override
-            protected void write(ByteBuffer header, ByteBuffer data) throws IOException {
-                LOGGER.finest(() -> "sending message of length " + ChunkHeader.length(ChunkHeader.peek(header)));
-                sendBinary(header, false);
-                sendBinary(data, true);
+            protected void write(ByteBuffer headerAndData) throws IOException {
+                // As in Engine.runWebSocket:
+                LOGGER.finest(() -> "sending message of length " + (headerAndData.remaining() - 2));
+                try {
+                    sendBinary(headerAndData).get(5, TimeUnit.MINUTES);
```

Comment:
Does this mean that any channel write operation should be under 5 minutes?

I know it's a bad practice, but what if a plugin is copying big files from the agent to the controller, that could easily take more than 5 minutes.

Reply:
You mean if you are using `archiveArtifacts` without `artifact-manager-s3`. Yes that could easily run over 5m for one _copy_ operation; but this timeout is for one transport frame, which is [currently only 8k](https://github.com/jenkinsci/remoting/blob/6f431148cd72051dc4b01559c038ef528fa3b424/src/main/java/hudson/remoting/AbstractByteBufferCommandTransport.java#L75-L78) (for WebSocket it could probably be increased considerably). Even one `Command` (broken into frames) would not necessarily be the full file size, since you would be using a `RemoteOutputStream` sending `ProxyOutputStream.Chunk`s which [default to 1Mb](https://github.com/jenkinsci/remoting/blob/6f431148cd72051dc4b01559c038ef528fa3b424/src/main/java/hudson/remoting/Channel.java#L1983-L1998). (all IIUC, but certainly worth testing)